### PR TITLE
New clear() method for Radio and Check buttons

### LINKED ISSUE #10922: ENH: clear() method for widgets.RadioButtons
### Enhancement proposed

Right now, RadioButtons can be created without activating any of the individual buttons, but they can not be cleared once created. It can be achieved in a dirty way, by hacking properties of the associated artists. However, I feel they need to come with a `clear()` method which deactivates previously selected one - it simply has to do two things: update the button colour and set value_selected to None. If this is acceptable, I'll be happy to send in a PR.

### PR DESCRIPTION
**Pull request by raamana**
## PR Summary

Continuing #10924 to clear up some merge/rebase issues and have a fresh start

Fixes #10922 

cc @jklymak 

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/widgets.py
Comment (jklymak) [inline lib/matplotlib/widgets.py]: We don't need to do "2." in python, and folks front on it here...

**Comment by jklymak** [comment]
Ha, I was just trying to get it moving again - I'm not actually familiar enough with the Radio Button stuff to review it.

**Comment by raamana** [comment]
Hi @jklymak, it was ALL approved in #10924 (3 approving reviews etc) except for waiting for the tests!! I hope we can merge this, so I can forget about it, and feel proud about having contributed to mighty matplotlib :)

**Comment by jklymak** [comment]
Hmmm, well, it was approved, but the tests never passed, hence it never was merged.  Its *still* not passing the tests so its still not going to be merged.  People often approve PRs assuming that the tests will catch anything unforeseen.  In this case, they appear to have, so you will have to pass that hurdle first.

**Comment by raamana** [comment]
I see. What do the tests in Azure pipeline test?
https://dev.azure.com/matplotlib/matplotlib/_build/results?buildId=1835&view=ms.vss-test-web.build-test-results-tab

what about this one from travis - no language set - are they not the same tests?
https://travis-ci.org/matplotlib/matplotlib/jobs/491327354

**Comment by raamana** [comment]
Also, all the errors seem to be with backed - see https://travis-ci.org/matplotlib/matplotlib/jobs/491327353

```
=================================== FAILURES ===================================
___________________ lib/matplotlib/tests/test_backend_qt.py ____________________
[gw0] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
Worker 'gw0' crashed while running 'lib/matplotlib/tests/test_backend_qt.py::test_fig_close[Qt5Agg]'
___________________ lib/matplotlib/tests/test_backend_qt.py ____________________
[gw1] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
Worker 'gw1' crashed while running 'lib/matplotlib/tests/test_backend_qt.py::test_fig_signals'
___________________ lib/matplotlib/tests/test_backend_qt.py ____________________
[gw3] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
Worker 'gw3' crashed while running 'lib/matplotlib/tests/test_backend_qt.py::test_correct_key[Qt5Agg-shift]'
___________________ lib/matplotlib/tests/test_backend_qt.py ____________________
[gw4] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
Worker 'gw4' crashed while running 'lib/matplotlib/tests/test_backend_qt.py::test_correct_key[Qt5Agg-control]'
___________________ lib/matplotlib/tests/test_backend_qt.py ____________________
[gw2] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
Worker 'gw2' crashed while running 'lib/matplotlib/tests/test_backend_qt.py::test_correct_key[Qt5Agg-lower]'
___________________ lib/matplotlib/tests/test_backend_qt.py ____________________
[gw5] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
Worker 'gw5' crashed while running 'lib/matplotlib/tests/test_backend_qt.py::test_correct_key[Qt5Agg-unicode_lower]'
___________________ lib/matplotlib/tests/test_backend_qt.py ____________________
[gw7] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
Worker 'gw7' crashed while running 'lib/matplotlib/tests/test_backend_qt.py::test_correct_key[Qt5Agg-control_alt]'
___________________ lib/matplotlib/tests/test_backend_qt.py ____________________
[gw8] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
Worker 'gw8' crashed while running 'lib/matplotlib/tests/test_backend_qt.py::test_correct_key[Qt5Agg-unicode_upper]'
___________________ lib/matplotlib/tests/test_backend_qt.py ____________________
[gw9] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
Worker 'gw9' crashed while running 'lib/matplotlib/tests/test_backend_qt.py::test_correct_key[Qt5Agg-backspace]'
_______________________ test_interactive_backend[qt5agg] _______________________
[gw6] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python
backend = 'qt5agg'
```

**Comment by raamana** [comment]
So when I run the tests locally, this time ensuring there is a DISPLAY via the VNC session, most of the test failures are gone. All the errors in the interactive backend are gone (esp the `webagg`, which I can't imagine why it would fail due to my code changes). What is left is one test in backend_pgf whatever that is!

```
test_backend_bases.py ........                                                                                                                                                                              [  7%]
test_backend_cairo.py s                                                                                                                                                                                     [  8%]
test_backend_nbagg.py .                                                                                                                                                                                     [  9%]
test_backend_pdf.py s............sssss.                                                                                                                                                                     [ 26%]
test_backend_pgf.py ssssssFss                                                                                                                                                                               [ 34%]
test_backend_ps.py .ss..s.s..sss                                                                                                                                                                            [ 46%]
test_backend_qt.py s..sssssssssss..............                                                                                                                                                             [ 72%]
test_backends_interactive.py ss.s.ss.                                                                                                                                                                       [ 79%]
test_backend_svg.py .s.ss.ss.ss                                                                                                                                                                             [ 89%]
test_backend_tools.py ...........  
```

**Comment by jklymak** [comment]
Looks like the qt5 errors are unrelated.

**Comment by raamana** [comment]
So the two errors remaining are to do with `rcparams` and LaTeX - see below. I think I will quit at this point and leave it to you guys to figure the errors out - I don't have anymore time for this right now unless it's something to do with code changes I made (or DIRECTLY related to it).

1.
```
E           ValueError: The following params are missing in the matplotlibrc.template file: dict_items([('boxplot.flierprops.markeredgewidth', [1.0, <function validate_float at 0x7f24739522f0>]), ('scatter.edgecolors', ['face', <function validate_string at 0x7f2473952f28>])])

test_rcparams.py:464:
```

2.
```
E           matplotlib.backends.backend_pgf.LatexError: LaTeX returned an error, probably missing font or error in preamble:
E           b'This is pdfTeXk, Version 3.141592-1.40.3 (Web2C 7.5.6)\n %&-line parsing enabled.\n**entering extended mode\n\n*! I can\'t find file `".ogut"\'.\n<to be read again> \n                   \\let \n<*> \\documentclass\n                  {minimal}\n(Press Enter to retry, or Control-D to exit)\nPlease type another input file name: ! I can\'t find file `".ogut"\'.\n<to be read again> \n                   \\let \n<*> \\documentclass\n                  {minimal}\n(Press Enter to retry, or Control-D to exit)\nPlease type another input file name: ! I can\'t find file `".ogut"\'.\n<to be read again> \n                   \\let \n<*> \\documentclass\n                  {minimal}\n(Press Enter to retry, or Control-D to exit)\nPlease type another input file name: ! I can\'t find file `\\begin{document}\'.\n<to be read again> \n                   \\let \n<*> \\documentclass\n                  {minimal}\n(Press Enter to retry, or Control-D to exit)\nPlease type another input file name: ! I can\'t find file `text\'.\n<to be read again> \n                   \\let \n<*> \\documentclass\n                  {minimal}\n(Press Enter to retry, or Control-D to exit)\nPlease type another input file name: ! I can\'t find file `\\typeout{pgf_backend_query_start}\'.\n<to be read again> \n                   \\let \n<*> \\documentclass\n                  {minimal}\n(Press Enter to retry, or Control-D to exit)\nPlease type another input file name: ! I can\'t find file `\\makeatletter\'.\n<to be read again> \n                   \\let \n<*> \\documentclass\n                  {minimal}\n(Press Enter to retry, or Control-D to exit)\nPlease type another input file name: ! I can\'t find file `\\@@end\'.\n<to be read again> \n                   \\let \n<*> \\documentclass\n                  {minimal}\n(Press Enter to retry, or Control-D to exit)\nPlease type another input file name: \n! Emergency stop.\n<to be read again> \n                   \\let \n<*> \\documentclass\n                  {minimal}\n!  ==> Fatal error occurred, no output PDF file produced!\nTranscript written on texput.log.\n'

../backends/backend_pgf.py:308: LatexError
---------------------------------------------------------------------------------------------- Captured stdout call -----------------------------------------------------------------------------------------------
This is pdfTeXk, Version 3.141592-1.40.3 (Web2C 7.5.6)
 %&-line parsing enabled.
**entering extended mode

*! I can't find file `".ogut"'.
<to be read again> 
                   \let 
<*> \documentclass
                  {minimal}
(Press Enter to retry, or Control-D to exit)
Please type another input file name: ! I can't find file `".ogut"'.
<to be read again> 
                   \let 
<*> \documentclass
                  {minimal}
(Press Enter to retry, or Control-D to exit)
Please type another input file name: ! I can't find file `".ogut"'.
<to be read again> 
                   \let 
<*> \documentclass
                  {minimal}
(Press Enter to retry, or Control-D to exit)
Please type another input file name: ! I can't find file `\begin{document}'.
<to be read again> 
                   \let 
<*> \documentclass
                  {minimal}
(Press Enter to retry, or Control-D to exit)
Please type another input file name: ! I can't find file `text'.
<to be read again> 
                   \let 
<*> \documentclass
                  {minimal}
(Press Enter to retry, or Control-D to exit)
Please type another input file name: ! I can't find file `\typeout{pgf_backend_query_start}'.
<to be read again> 
                   \let 
<*> \documentclass
                  {minimal}
(Press Enter to retry, or Control-D to exit)
Please type another input file name: ! I can't find file `\makeatletter'.
<to be read again> 
                   \let 
<*> \documentclass
                  {minimal}
(Press Enter to retry, or Control-D to exit)
Please type another input file name: ! I can't find file `\@@end'.
<to be read again> 
                   \let 
<*> \documentclass
                  {minimal}
(Press Enter to retry, or Control-D to exit)
Please type another input file name: 
! Emergency stop.
<to be read again> 
                   \let 
<*> \documentclass
                  {minimal}
!  ==> Fatal error occurred, no output PDF file produced!
Transcript written on texput.log.
```

**Comment by raamana** [comment]
Yes @jklymak, these errors seem to be unrelated to my work. What's the process for accepting PRs in a situation like this here?

**Comment by jklymak** [comment]
I'll try recycling...

**Comment by anntzer** [comment]
Some (real, this-PR-related) style issues: https://travis-ci.org/matplotlib/matplotlib/jobs/492794388.

**Comment by raamana** [comment]
thanks - fixed the two flake issues:
```
(py36) $ 13:28:22 SQuark-3 matplotlib >>  flake8 --statistics widgets.py
(py36) $ 13:28:35 SQuark-3 matplotlib >>
```

**Comment by raamana** [comment]
Lovely - great to see it all green here :)

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py]: Please keep numpydoc-style.

File: lib/matplotlib/widgets.py
Comment (raamana) [reply lib/matplotlib/widgets.py]: Sure - I suggest these kind of checks be made part of flake test.

File: lib/matplotlib/widgets.py
Comment (raamana) [reply lib/matplotlib/widgets.py]: Sure.. Also, I've been working on this for over a year now - I don't remember making this change. In fact, this does not seem like the type of code I usually write.

File: lib/matplotlib/widgets.py
Comment (timhoffm) [reply lib/matplotlib/widgets.py]: I don't think there is an error code for too short lines.

**Comment by raamana** [comment]
Style comments are addressed.

File: lib/matplotlib/widgets.py
Comment (timhoffm) [reply lib/matplotlib/widgets.py]: The question is still open (only marked as outdated because the variable got renamed?

**Comment by raamana** [comment]
Ok, I think I am done. This has been a great learning exercise trying to contribute to a complex OS library :). I totally severely underestimated what it takes :)

**Comment by raamana** [comment]
This also closes #10924

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py]: Please leave as is. You mustn't remove the default value. Also, renaming a parameter is a backward-incompatible API change. That would need a deprecation period. I think  that's not worth the effort, but if you really want to change it, add a deprecation note to next_api_changes. Since recently, we have a decorator that can handle renaming of parameters with deprecation logic, which would have to be added to the code.

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py]: `ax` description is lost. The listed parameters must be complete.

Also, I think the original parameter descriptions are better, so the whole docstring should essentially stay as is -- except maybe fot the types, for which we use `list of str` and not `List[str]` according to numpydoc conventions.

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py]: That's a bit misleading: toggling is not controlled via a bool state.

I would remove this sentence and add the information to the state parameter. See below.

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py]: Suggestion:
~~~
       state : bool, optional
            Determines the target state of the button. Boolean values
            set the state explicitly. If not given, the state gets toggled.

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py]:
```suggestion
        """Return a list of labels currently checked by user."""
```

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py]: I find the original description more to the point. This method connects a callback function. What you are describing is the execution of the callback. If you feel the need that this should be explained in the docstring, that should go in a second paragraph after "Connect the callback ..." (optional to me).

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py]:
```suggestion
    Connect to the RadioButtons with the :meth:`on_clicked` method.
```

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py]: See https://matplotlib.org/devdocs/api/_as_gen/matplotlib.artist.Artist.add_callback.html?highlight=signature for a good way to describe a callable.

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py]:
```suggestion
            A connection id, which can be used to disconnect.
```

**Comment by raamana** [comment]
Thanks Tim for detailed review and comments.

File: lib/matplotlib/widgets.py
Comment (raamana) [reply lib/matplotlib/widgets.py]: fixed it, thanks.

File: lib/matplotlib/widgets.py
Comment (raamana) [reply lib/matplotlib/widgets.py]: Good pointer, used it to improve docs.

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py]: Please make really sure you adhere to the documentation conventions. That includes indentation in docstrings, on which ReST/sphinx/numpydoc is very strict.

I know it's cumbersome, but we need a certain quality for PRs to be merged; and I don't have the time to comment on every formatting issue.

At best, build the docs locally and check the result. Alternatively, check the documentation built via CircleCI
grafik
grafik

You'll see that the formatting is messed because the indentations here are all not right.

https://18374-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/api/widgets_api.html?highlight=widgets#matplotlib.widgets.CheckButtons

Also, if you convert a docstring from the old format to numpydoc, I expect the conversion to be complete, i.e. add types, no parameter formatting etc.

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py]:
```suggestion
     rectangles
```

**Comment by QuLogic** [comment]
I rebased, cleaned up the conflicts, and squashed to one commit. The `clear` method on `RadioButtons` does allow setting _no_ buttons selected, which is something that's no longer allowed in `__init__`: https://github.com/matplotlib/matplotlib/commit/3c306a3e0762c92e50522e982dd95664888aadaa#r106438132

I wonder if instead, we want to remember the initial active index set in `__init__` and make `clear` reset to that?

**Comment by jklymak** [comment]
I think I agree that radio buttons should always have at least one element selected.   If you need a "none of the above", then that should be another button choice.  https://uxplanet.org/radio-buttons-ux-design-588e5c0a50dc

**Comment by jklymak** [comment]
I'll move to draft

**Comment by QuLogic** [comment]
OK, `RadioButtons.clear` now resets back to the original setup.

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py:1904]: I'm wondering whether this should be named `reset()`.

Pro `reset`: This is the correct terminology. You cannot clear a radio button group.
Pro `clear`: Consistency with other widget. There are a number of widgets with `clear` methods, but none with `reset`.

**Review by timhoffm — COMMENTED** [review]
This needs a what's new entry.

File: lib/matplotlib/tests/test_widgets.py
Comment (QuLogic) [reply lib/matplotlib/tests/test_widgets.py]: Done.

**Review by jklymak — APPROVED** [review]

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py:1203]:
```suggestion
        Modify the state of a check button by index.
```

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py:1214]:
```suggestion
            If a boolean value, set the state explicitly. If no value is
            provided, the state is toggled.
```

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py:1273]:
```suggestion
        """Uncheck all checkboxes."""
```

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py:1905]:
```suggestion
        """Reset the active button to the initially active one."""
```

**Review by timhoffm — COMMENTED** [review]
Minor docstring improvements.

**Review by timhoffm — APPROVED** [review]

**Pull request merged**
